Library/Homebrew/cask/lib/hbc/installer.rb

```diff
@@ -420,10 +443,10 @@ def purge_versioned_files
         end
       end
       @cask.metadata_versioned_path.rmdir_if_possible
-      @cask.metadata_master_container_path.rmdir_if_possible
+      @cask.metadata_master_container_path.rmdir_if_possible unless upgrade?
```

Comment:
Is this check here necessary, since this is not `rm_rf` but only `rmdir`?

Reply:
Yes. Line 445 clears the Cask's versioned metadata (which is what we need for `finalize_upgrade`), but 446 below deletes the whole metadata folder (that is, `$(brew --prefix)/Caskroom/<cask_name>/`, which we only need for `uninstall`).